File: Demo/CrossDK-demo Objective-C/CrossDK-demo Objective-C/CrossDKViewController/CrossDKViewController.m
//
//  CrossDKViewController.m
//  CrossDK-demo Objective-C
//
//  Created by Adikteev on 09/11/2021.
//

#import "CrossDKViewController.h"

@interface CrossDKViewController () <CrossDKOverlayDelegate>
    @property (strong, nonatomic) CrossDKOverlay* crossDKOverlay;
@end

@implementation CrossDKViewController

// MARK: - viewDidLoad

- (void)viewDidLoad {
    [super viewDidLoad];

    _crossDKOverlay = [[CrossDKOverlay alloc] init];
    _crossDKOverlay.delegate = self;
    for (UIButton *overlayButton in self.overlayButtons) {
        overlayButton.layer.cornerRadius = 18;
    }
}

/// Banner overlay button clicked.
- (IBAction)bannerOverlayButtonPressed:(UIButton *)sender {
    UIWindow* window = self.view.window;
    if (window != nil) {
        [_crossDKOverlay displayWithWindow:window format:OverlayFormatBanner position:OverlayPositionBottom withCloseButton:true];
    }
}

/// Mid size overlay button clicked.
- (IBAction)midSizeOverlayButtonPressed:(UIButton *)sender {
    UIWindow* window = self.view.window;
    if (window != nil) {
        [_crossDKOverlay displayWithWindow:window format:OverlayFormatMid_size position:OverlayPositionBottom withCloseButton:true];
    }
}

/// Interstitial overlay button clicked.
- (IBAction)interstitialOverlayButtonPressed:(UIButton *)sender {
    UIWindow* window = self.view.window;
    if (window != nil) {
        [_crossDKOverlay displayWithWindow:window format:OverlayFormatInterstitial position:OverlayPositionBottom withCloseButton:true];
    }
}

// MARK: - CrossDKOverlayDelegate

- (void)overlayWillStartPresentation {
    NSLog(@"Overlay will start presentation");
}

- (void)overlayDidFinishPresentation {
    NSLog(@"Overlay did finish presentation");
}

- (void)overlayWillStartDismissal {
    NSLog(@"Overlay will start dismissal");
}

- (void)overlayDidFinishDismissal {
    NSLog(@"Overlay did finish dismissal");
}

- (void)overlayStartsPlayingVideo {
    NSLog(@"Overlay starts playingVideo");
}

- (void)overlayPlayedHalfVideo {
    NSLog(@"Overlay played half video");
}

- (void)overlayDidFinishPlayingVideo {
    NSLog(@"Video Overlay did finish playing video");
}

- (void)overlayShowsRecommendedAppInAppStore {
    NSLog(@"Overlay shows recommended app in AppStore");
}

- (void)overlayDidFailToLoadWithError:(NSError * _Nonnull)error {
    NSLog(@"Overlay did fail to load with error : %ld", (long)error);
}

- (void)overlayUnavailableWithError:(enum OverlayError)error {
    switch(error) {
        case OverlayErrorUnsupportedOSVersion:
            NSLog(@"Overlay error: unsupported iOS Version");
            break;
        case OverlayErrorUnavailableWindowScene:
            NSLog(@"Overlay error: unavailable window scene");
            break;
        case OverlayErrorUnavailableRecommendation:
            NSLog(@"Overlay error: unavailable recommendation");
            break;
        case OverlayErrorNoConfiguration:
            NSLog(@"Overlay error: no configuration");
            break;
    }
}

@end

File: Demo/CrossDK-demo UIKit/CrossDK-demo UIKit/CrossDKViewController/CrossDKViewController.swift
//
//  CrossDKViewController.swift
//  CrossDK-demo UIKit
//
//  Created by Adikteev on 19/07/2021.
//

import UIKit
import CrossDK

class CrossDKViewController: UIViewController {

    // MARK: - Outlets

    @IBOutlet private var overlayButtons: [UIButton]!

    // MARK: - Private properties

    private let crossDKOverlay = CrossDKOverlay()
    private let systemVersion = UIDevice.current.systemVersion

    // MARK: - Override Funcs

    override func viewDidLoad() {
        super.viewDidLoad()
        
        crossDKOverlay.delegate = self
        overlayButtons.forEach { $0.layer.cornerRadius = 18 }
    }
}

// MARK: - Private Actions

private extension CrossDKViewController {
    /// Overlay button clicked.
    @IBAction func overlayButtonPressed(_ sender: UIButton) {
        guard let window = view.window else { return }

        switch sender {
        case overlayButtons[0]:
            crossDKOverlay.display(window: window, format: .banner, position: .bottom, withCloseButton: true)
        case overlayButtons[1]:
            crossDKOverlay.display(window: window, format: .mid_size, position: .bottom, withCloseButton: true)
        case overlayButtons[2]:
            crossDKOverlay.display(window: window, format: .interstitial, position: .bottom, withCloseButton: true)
        default:
            break
        }
    }
}

extension CrossDKViewController: CrossDKOverlayDelegate {
    func overlayWillStartPresentation() {
        NSLog("Overlay will start presentation")
    }

    func overlayDidFinishPresentation() {
        NSLog("Overlay did finish presentation")
    }

    func overlayWillStartDismissal() {
        NSLog("Overlay will start dismissal")
    }

    func overlayDidFinishDismissal() {
        NSLog("Overlay did finish dismissal")
    }

    func overlayStartsPlayingVideo() {
        NSLog("Overlay starts playing video")
    }

    func overlayPlayedHalfVideo() {
        NSLog("Overlay played half video")
    }

    func overlayDidFinishPlayingVideo() {
        NSLog("Overlay did finish playing video")
    }

    func overlayShowsRecommendedAppInAppStore() {
        NSLog("Overlay shows recommended app in App Store")
    }

    func overlayDidFailToLoad(error: Error) {
        NSLog("Overlay did fail to load with error: \(error)")
    }

    func overlayUnavailable(error: CrossDKOverlay.OverlayError) {
        switch error {
        case .unsupportedOSVersion:
            NSLog("iOS 14 is required to run CrossDK. Current iOS version is: \(systemVersion)")
        case .unavailableWindowScene:
            NSLog("Unavailable WindowScene")
        case .unavailableRecommendation:
            NSLog("Unavailable Recommendation")
        case .noConfiguration:
            NSLog("No configuration WindowScene")
        @unknown default:
            NSLog("Overlay unavailable")
        }
    }
}
